p2p/transport/pipe/pipetransport.go

```diff
@@ -0,0 +1,93 @@
+package pipetransport
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"sync"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+	tpt "github.com/libp2p/go-libp2p-transport"
+	tptu "github.com/libp2p/go-libp2p-transport-upgrader"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type PipeTransport struct {
+	mlistenchans *sync.RWMutex
+	listenchans  map[string]chan *PipeConn
+
+	upgrader *tptu.Upgrader
+}
+
+var _ tpt.Transport = (*PipeTransport)(nil)
+
+func New(u *tptu.Upgrader) *PipeTransport {
+	return &PipeTransport{
+		mlistenchans: new(sync.RWMutex),
+		listenchans:  make(map[string]chan *PipeConn),
+		upgrader:     u,
+	}
+}
+
+func (t *PipeTransport) closeListener(addr string) {
+	t.mlistenchans.Lock()
+	defer t.mlistenchans.Unlock()
+
+	ch, ok := t.listenchans[addr]
+	if !ok {
+		return
+	}
+	close(ch)
+	delete(t.listenchans, addr)
+}
+
+func (t *PipeTransport) CanDial(addr ma.Multiaddr) bool {
+	protocols := addr.Protocols()
+	return len(protocols) == 1 && protocols[0].Code == ma.P_P2P
```

Comment:
This filter is too wide I think, it will also catch arbitrary `/p2p/QmTarget` addresses which almost certainly cannot be dialed with the pipe.
Maybe we should check the peer Id matching our own here?

Reply:
i wanted to do this, but we wont always know the peer ID at the time of creating the transport, so i left it blank. do we think we should lock it in?

Reply:
Hrm, tricky. My worry is that we'll start trying to pipe-dial arbitrary addrs (I've seen naked `/p2p` addrs in the wild).

Reply:
We should always know our own peer ID when creating the transport. You can define the constructor to accept it as a parameter (instead of accepting an upgrader).